x/twap/listeners_test.go

```diff
@@ -3,35 +3,99 @@ package twap_test
-func (s *TestSuite) TestCreateTwoAssetPoolFlow() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-
-	expectedTwap, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-
-	twap, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedTwap, twap)
-
-	twap, err = s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedTwap, twap)
-}
-
-// TODO: Write test for create pool, swap, and EndBlock in same block, we use post-swap spot price in record.
-
-// Tests that after a swap, we are triggering internal tracking logic for a pool.
-func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-	s.BeginNewBlock(false)
-	s.RunBasicSwap(poolId)
-
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	const (
+		Uni2               = 0
+		Balancer3AssetPool = 1
+	)
+
+	tests := []struct {
+		name     string
+		poolType int
+		// if this field is set true, we swap in the same block with pool creation
+		runSwap bool
+	}{
+		{
+			"Uni2 Pool, no swap on pool creation block",
+			Uni2,
+			false,
+		},
+		{
+			"Uni2 Pool, swap on pool creation block",
+			Uni2,
+			true,
+		},
+		{
+			"Three asset balancer pool, no swap on pool creation block",
+			Balancer3AssetPool,
+			false,
+		},
+		{
+			"Three asset balancer pool, no swap on pool creation block",
+			Balancer3AssetPool,
+			true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.SetupTest()
+		s.Run(tc.name, func() {
+			var poolId uint64
+			// prepare pool according to test case
+			if tc.poolType == Uni2 {
+				poolId = s.PrepareUni2PoolWithCoins(defaultUniV2Coins[0], defaultUniV2Coins[1])
+			} else {
+				poolId = s.PrepareBalancerPoolWithCoins(defaultThreeAssetCoins...)
+			}
+
+			if tc.runSwap {
+				s.RunBasicSwap(poolId)
+			}
+
+			timeBeforeBeginBlock := s.Ctx.BlockTime()
+			s.twapkeeper.EndBlock(s.Ctx)
+			s.BeginNewBlock(false)
+
+			// check that creating a pool saved a new state of changed pools
+			s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+
+			// check that all twap records have been created for all denom pairs in pool
+			twapRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denomPairs0, _ := types.GetAllUniqueDenomPairs(denoms)
```

Comment:
What is the significance of the 0 in `denomPairs0`?

Reply:
`GetAllUniqueDenomPairs` would return two slices of string, the first slice indicating the first asset of the pair, the second slice indicating the second asset of the pair. We would use both in normal cases, denomPairs0[1], denomParis1[1] or something similar